hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/CompactHelpers.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.hudi.table.action.HoodieWriteMetadata;
+
+import org.apache.avro.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Base class helps to perform compact.
+ *
+ * @param <T> Type of payload in {@link org.apache.hudi.common.model.HoodieRecord}
+ */
+public class CompactHelpers<T extends HoodieRecordPayload> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CompactHelpers.class);
+
+  private CompactHelpers() {
+  }
+
+  private static class CompactHelperHolder {
+    private static final CompactHelpers COMPACT_HELPERS = new CompactHelpers();
+  }
+
+  public static CompactHelpers newInstance() {
+    return CompactHelperHolder.COMPACT_HELPERS;
+  }
+
+  public HoodieWriteMetadata<HoodieData<WriteStatus>> compact(
+      HoodieEngineContext context, HoodieTable table,
+      HoodieCopyOnWriteTableOperation copyOnWriteTableOperation, HoodieWriteConfig config,
+      String compactionInstantTime, AbstractHoodieWriteClient writeClient,
+      TaskContextSupplier taskContextSupplier) {
+    HoodieTimeline pendingCompactionTimeline = table.getActiveTimeline().filterPendingCompactionTimeline();
+    HoodieInstant inflightInstant = HoodieTimeline.getCompactionInflightInstant(compactionInstantTime);
+    if (pendingCompactionTimeline.containsInstant(inflightInstant)) {
+      writeClient.rollbackInflightCompaction(inflightInstant, table);
+      table.getMetaClient().reloadActiveTimeline();
+    }
+
+    HoodieWriteMetadata<HoodieData<WriteStatus>> compactionMetadata = new HoodieWriteMetadata<>();
+    compactionMetadata.setWriteStatuses(context.createEmptyHoodieData());
+    try {
+      // generate compaction plan
+      // should support configurable commit metadata
+      HoodieCompactionPlan compactionPlan = CompactionUtils.getCompactionPlan(
+          table.getMetaClient(), compactionInstantTime);
+
+      if (compactionPlan == null || (compactionPlan.getOperations() == null)
+          || (compactionPlan.getOperations().isEmpty())) {
+        // do nothing.
+        LOG.info("No compaction plan for instant " + compactionInstantTime);
+        return compactionMetadata;
+      }
+
+      HoodieInstant instant = HoodieTimeline.getCompactionRequestedInstant(compactionInstantTime);
+      // Mark instant as compaction inflight
+      table.getActiveTimeline().transitionCompactionRequestedToInflight(instant);
+      table.getMetaClient().reloadActiveTimeline();
+
+      HoodieTableMetaClient metaClient = table.getMetaClient();
+      TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
+
+      // Here we firstly use the table schema as the reader schema to read
+      // log file.That is because in the case of MergeInto, the config.getSchema may not
+      // the same with the table schema.
+      try {
```

Comment:
I guess these are not invoked in flink engine? can you double check that please. HoodieFlinkMergeOnReadTableCompactor

Reply:
@danny0405 : need some help on  understanding some behavior diff between spark and flink wrt compation. 
there is a code block which does some schema inference in spark which is skipped in flink. 
[Flink link](https://github.com/apache/hudi/blob/47ed91799943271f219419cf209793a98b3f09b5/hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/table/action/compact/HoodieFlinkMergeOnReadTableCompactor.java#L88), [spark link](https://github.com/apache/hudi/blob/47ed91799943271f219419cf209793a98b3f09b5/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/compact/HoodieSparkMergeOnReadTableCompactor.java#L90).

In other words, this below code block is missing in flink. 
```
 HoodieTableMetaClient metaClient = hoodieTable.getMetaClient();
    TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);

    // Here we firstly use the table schema as the reader schema to read
    // log file.That is because in the case of MergeInto, the config.getSchema may not
    // the same with the table schema.
    try {
      Schema readerSchema = schemaUtil.getTableAvroSchema(false);
      config.setSchema(readerSchema.toString());
    } catch (Exception e) {
      // If there is no commit in the table, just ignore the exception.
    }
```
 Is this intentional? can you help us understand the behavior difference.